---
int32 mode
int32 res